ow_power_system/src/PrognoserInputHandler.cpp

```diff
@@ -0,0 +1,218 @@
+    ROS_ERROR("Failed to load PrognoserInputHandler system config!");
+    return false;
+  }
+
+  return true;
+}
+
+/*
+ * Called from Initialize(). Reads the system.cfg file within ow_power_system
+ * and stores its information in relevant class variables.
+ */
+bool PrognoserInputHandler::loadSystemConfig()
+{
+  double loop_rate;
+  try
+  {
+    auto system_config_path = ros::package::getPath("ow_power_system")
+      + "/config/system.cfg";
+    auto system_config = ConfigMap(system_config_path);
+    m_base_voltage = system_config.getDouble("base_voltage");
+    m_voltage_range = system_config.getDouble("voltage_range");
+    m_min_temperature = system_config.getDouble("min_temperature");
+    m_max_temperature = system_config.getDouble("max_temperature");
+    m_battery_lifetime = system_config.getDouble("battery_lifetime");
+    m_efficiency = system_config.getDouble("efficiency");
+    m_temperature_dist = uniform_real_distribution<double>(m_min_temperature,
+                                                          m_max_temperature);
+    m_baseline_wattage = system_config.getDouble("baseline_power");
+    m_max_gsap_input_watts = system_config.getDouble("max_gsap_power_input");
+    m_time_interval = 1 / (system_config.getDouble("loop_rate"));
+  }
+  catch(const std::exception& err)
+  {
+    ROS_ERROR_STREAM("OW_POWER_SYSTEM ERROR: " << err.what() 
+                     << " while attempting to read system.cfg"
+                     << " in an input handler!\n"
+                     << "Ensure system.cfg is formatted properly before "
+                     << "restarting the simulation.");
+    return false;
+  }
+
+  return true;
+}
+
+void PrognoserInputHandler::applyMechanicalPower(double mechanical_power,
+                                                 bool process_lock)
+{
+  // Set the current unprocessed mechanical power to the most recent value
+  // from jointStatesCb.
+  m_unprocessed_mechanical_power = mechanical_power;
+
+  if ((!process_lock) && (!m_processing_power_batch))
+  {
+    m_mechanical_power_to_be_processed = m_unprocessed_mechanical_power;
+    m_unprocessed_mechanical_power = 0.0;
+    m_trigger_processing_new_power_batch = true;
+  }
+}
+
+/*
+ * Returns a random value between the lower and upper bounds of the random
+ * generator, to mimic a sensor's temperature measurement.
+ */
+double PrognoserInputHandler::generateTemperatureEstimate()
+{
+  return m_temperature_dist(m_random_generator);
+}
+
+/*
+ * Calculates a voltage value with some variance to be used as a data input. Voltage
+ * values will slowly drop over time.
+ */
+double PrognoserInputHandler::generateVoltageEstimate()
+{
+  // Create voltage estimate with pseudorandom noise generator - needs to decrease over time
+  double timelapse = ros::Time::now().toSec() - m_init_time;            // s
+  double min_V = m_base_voltage + (m_battery_lifetime - timelapse)      // [V]
+                 / m_battery_lifetime * 0.8;
```

Comment:
I think we brushed over this in today's Dev Day, but essentially, we're just modeling a basic voltage sensor whose values aren't actually used for any predictions; it's just there to get a general idea of what the voltage of the power system would theoretically be as time goes on. The function itself just calculates a voltage range based on how long the simulation has been running, and then pulls a random value from that range. The size of the range is constant but the min & max go down over time.

I can add more details in a comment in the function though, it's warranted here. I didn't even know until today's meeting that the voltage/temperature values are completely unused by GSAP after initialization.

Reply:
I think this will get removed following our discussion in the dev day this morning, correct?

Reply:
I think we decided to leave this and the temperature generation functions in, even though they aren't used for anything at this point, simply because they technically mimic a sensor and could be used in the future.

For what it's worth, even though GSAP does not use voltage & temperature inputs after the initial setup, it still requires them to be sent in as part of the input setup (at least to my understanding).